### DISCUSSION TITLE
Seeing a weird exception after a successful login

### DISCUSSION THREAD

**Discussion by stevesobol** [quarkusio/quarkus#43719] [Q&A] [answered]
Using the latest Quarkus on graalvm-jdk-21.0.4+8.1. Windows 11 23H2, 64-bit, with all the updates, if it matters.

I'm working on a custom SecurityIdentityProvider. I'm trying to get it working with form authentication. Initially I'm just trying to get it working with a predefined username (bob) and password (bob).

```java
package com.example;

import io.quarkus.security.AuthenticationFailedException;
import io.quarkus.security.identity.AuthenticationRequestContext;
import io.quarkus.security.identity.IdentityProvider;
import io.quarkus.security.identity.SecurityIdentity;
import io.quarkus.security.identity.request.UsernamePasswordAuthenticationRequest;
import io.quarkus.security.runtime.QuarkusPrincipal;
import io.quarkus.security.runtime.QuarkusSecurityIdentity;
import io.smallrye.mutiny.Uni;
import jakarta.enterprise.context.ApplicationScoped;

import java.util.Map;

@ApplicationScoped
public class ExampleIdentityProvider implements IdentityProvider<UsernamePasswordAuthenticationRequest> {
    private static final Map<String, String> CREDENTIALS = Map.of("bob", "bob");

    @Override
    public Class<UsernamePasswordAuthenticationRequest> getRequestType() {
        return UsernamePasswordAuthenticationRequest.class;
    }

    @Override
    public Uni<SecurityIdentity> authenticate(UsernamePasswordAuthenticationRequest request,
                                              AuthenticationRequestContext authenticationRequestContext) {
        if (new String(request.getPassword().getPassword()).equals(CREDENTIALS.get(request.getUsername()))) {
			return Uni.createFrom().item(QuarkusSecurityIdentity.builder()
				.setPrincipal(new QuarkusPrincipal(request.getUsername()))
                    .addCredential(request.getPassword())
                    .setAnonymous(false)
                    .addRole("admin")
                    .build());
        }
        throw new AuthenticationFailedException("password invalid or user not found");
    }
}
```

`application.properties` looks like this:

```properties
quarkus.package.jar.type=uber-jar
quarkus.quinoa.ui-root-path=app
quarkus.quinoa.enable-spa-routing=true

quarkus.http.auth.form.enabled=true
quarkus.http.auth.form.login-page=login
quarkus.http.auth.form.landing-page=success
quarkus.http.auth.form.error-page=error

quarkus.datasource.db-kind=mariadb
quarkus.datasource.username=db-username
quarkus.datasource.password=db-password
quarkus.datasource.jdbc.url=jdbc:mariadb://localhost/somedatabase
```

Although I am using quinoa for part of the application, the login, success and error pages are standard Qute templates.

I can log in successfully. But instead of seeing the "success" page, I get an exception

```
java.lang.IllegalArgumentException: No IdentityProviders were registered to handle AuthenticationRequest io.quarkus.security.identity.request.TrustedAuthenticationRequest@326d3217
	at io.quarkus.security.runtime.QuarkusIdentityProviderManagerImpl.authenticate(QuarkusIdentityProviderManagerImpl.java:61)
	at io.quarkus.security.identity.IdentityProviderManagerCreator_ProducerMethod_ipm_kfECvhsqeBIW24qB5qtLmxqE8Dw_ClientProxy.authenticate(Unknown Source)
	at io.quarkus.vertx.http.runtime.security.FormAuthenticationMechanism.authenticate(FormAuthenticationMechanism.java:266)
	at io.quarkus.vertx.http.runtime.security.HttpAuthenticator.createSecurityIdentity(HttpAuthenticator.java:223)
	at io.quarkus.vertx.http.runtime.security.HttpAuthenticator.attemptAuthentication(HttpAuthenticator.java:181)
	at io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder$AuthenticationHandler.handle(HttpSecurityRecorder.java:284)
	at io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder$AuthenticationHandler.handle(HttpSecurityRecorder.java:233)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1285)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:177)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:140)
	at io.quarkus.vertx.http.runtime.devmode.VertxHttpHotReplacementSetup.handleHotReplacementRequest(VertxHttpHotReplacementSetup.java:144)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$HotReplacementRoutingContextHandler.handle(VertxHttpRecorder.java:1652)
```

(many lines of stacktrace omitted)

`/restricted` points to this method:

```java
	@GET
	@RolesAllowed("admin")
	@Path("/restricted")
	public String me(@Context SecurityContext securityContext) {
		return securityContext.getUserPrincipal().getName();
	}
```

Any ideas? What am I doing wrong?

**Comment by stevesobol** [comment]
Incidentally, I know the login is successful. I had a couple lines of code placed before the `return` in `authenticate()` and when I logged in successfully, I saw the log entries. The problem seems to occur when my web browser requests `/restricted` after a successful login.

**Reply by sberyozkin** [reply]
@stevesobol I see that `FormAuthenticationMechanism` initiates `UsernamePasswordAuthenticationRequest` during the initial login [here](https://github.com/quarkusio/quarkus/blob/main/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/FormAuthenticationMechanism.java#L155), and then `TrustedAuthenticationRequest` when the already authenticated cookie is available [here](https://github.com/quarkusio/quarkus/blob/main/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/FormAuthenticationMechanism.java#L268).

Do you set `@Alternative` on your custom provider ? If  yes then your `UsernamePasswordAuthenticationRequest` aware provider will be the only identity provider that is checked, thus the `TrustedAuthenticationRequest` is lost. 

But I guess, if you don't use `@Alternative`, then your provider will compete with the default  `UsernamePasswordAuthenticationRequest` aware provider. But I believe this is what you should do, only make your provider higher priority, with the `TrustedAuthenticationRequest` provider being also available.

**Reply by stevesobol** [reply]
Didn't work. But thank you for the information. I've dug through the source code trying to figure out what's happening, and as far as I can tell, making it an injectable bean _should_ work. I tried both `@Priority(1)` and `@Priority(100)`. No love.

**Reply by stevesobol** [reply]
But I still only have the one class. I don't have anything that implements `IdentityProvider<TrustedAuthenticationRequest>`.

**Reply by stevesobol** [reply]
Solution: Create another class that handles `TrustedAuthenticationRequests` - which I need to do anyhow. The `@RolesAllowed` controller annotation requires me to assign roles in this class's `authenticate()` method:

```java
@ApplicationScoped
public class ExampleIdentityProvider2 implements IdentityProvider<TrustedAuthenticationRequest> {

    @Override
    public Class<TrustedAuthenticationRequest> getRequestType() {
        return TrustedAuthenticationRequest.class;
    }

    @Override
    public Uni<SecurityIdentity> authenticate(TrustedAuthenticationRequest request,
                                              AuthenticationRequestContext authenticationRequestContext) {
		if (request.getPrincipal() != null) {
			return Uni.createFrom().item(QuarkusSecurityIdentity.builder()
				.setPrincipal(new QuarkusPrincipal(request.getPrincipal()))
				.addRoles(Set.of("admin", "create", "read", "update", "delete"))
					.addAttribute("zxc","qwe")
				.build());
		}

		throw new AuthenticationFailedException("password invalid or user not found");
    }
}
```

So, basically, the class I originally created handles _**authentication**_, and this class handles _**authorization**_.

@sberyozkin - I owe you a beer, or other beverage of your choosing. Understanding that there are actually two requests that need to be processed was the key to fixing my problem. Thank you very much!

**Reply by sberyozkin** [reply]
@stevesobol Np at all,

>So, basically, the class I originally created handles authentication, and this class handles authorization.

It does not have to be done in the 2nd identity provider,  a custom `SecurityIdentityAugmentor` is where one would typically augment the identity.

You should still not be required to register the 2nd provider though if you don't use `@Alternative`. FYI, you should have 1002 or higher priority for the 1st, user name/password aware provider: https://github.com/quarkusio/quarkus-security/blob/main/src/main/java/io/quarkus/security/identity/IdentityProvider.java#L57

Create a reproducer please if you can't make it work, thanks

**Reply by stevesobol** [reply]
Using a `SecurityIdentityAugmentor` and the original identity provider doesn't work, even after adding the `@Priority(1002)` annotation to the original identity provider. I'm going to create a clean project and try this again from scratch and see if I can get it to work, and if I can't, I'll create a repo, push the code there and post the URL here.

**Reply by stevesobol** [reply]
OK. Starting fresh, using quarkus-security, Qute and Renarde, I still get the same exception. Reproducer is here:

https://github.com/lobosstudios/broken-quarkus-auth-poc

**Reply by stevesobol** [reply]
Any thoughts? Anyone? 😳

**Reply by sberyozkin** [reply]
@stevesobol The problem we were trying to solve here was how to make sure that form authentication applications can only replace only one of the two required identity providers, the one managing name and password,  if it has to be customised. 

Your reproducer is about something else, why, in case of the wrong password, user is redirected to the default home page...

I guess an error page must be configured

**Reply by sberyozkin** [reply]
I see it is configured,  @michalvavrik , can you please have a quick look next week?

**Reply by michalvavrik** [reply]
Sure, I'll check.

**Reply by mkouba** [reply, marked as answer]
AFAIK you always need to implement both `IdentityProvider<UsernamePasswordAuthenticationRequest>` and `IdentityProvider<TrustedAuthenticationRequest>` if using form-based authentication. At least that's what `quarkus-security-jpa` generates and I also used it successfully with my apps. The `IdentityProvider<TrustedAuthenticationRequest>` implementation is used to authenticate request from an encrypted cookie.

In other words, if an `IdentityProvider<UsernamePasswordAuthenticationRequest>` succeeds the authentication information is stored in an encrypted cookie which is sent with a follow-up requests and for these requests the `IdentityProvider<TrustedAuthenticationRequest>`  is used to build a `SecurityIdentity`.

**Reply by sberyozkin** [reply]
Hi @mkouba Thanks, sure, but if we only register a higher priority `IdentityProvider<UsernamePasswordAuthenticationRequest>` which is preferred to the built-in `IdentityProvider<UsernamePasswordAuthenticationRequest>`, the built-in `IdentityProvider<TrustedAuthenticationRequest>` should still be available ?

**Reply by mkouba** [reply]
> the built-in IdentityProvider<TrustedAuthenticationRequest> should still be available ?

Which built-in `IdentityProvider<TrustedAuthenticationRequest>` do you have in mind? The reproducer app does not seem to include `quarkus-security-jpa`, `quarkus-security-webauthn` or any other extension that would provide such implementation?

**Reply by sberyozkin** [reply]
@mkouba Oh yes, this must be it, I was under the impression only the first one can to be customized, but if the one for the `TrustedAuthenticationRequest` depends on the same database table for ex, then it does not really make sense to customize a single one independently of the other one.

@stevesobol Sorry, I was thinking that you may want to customise some username and password preprocessing, assuming the default provider for the `TrustedAuthenticationRequest`  can work with the same data your custom username and password identity provider can. If it is still the case then, as @mkouba pointed out, you'd need to use an extension like `quarkus-security-jpa`. Otherwise, indeed, both type of providers have to be customized for the form authentication to work. Sorry for the confusion if any

**Comment by phillip-kruger** [comment]
@sberyozkin ^^^
